File: python/src/dashgps/registry.py
"""Format registry and auto-detection. spec/00-model.md.

Sniffing is budgeted by declared IO cost. That is what stops one non-matching file in a batch of
1,200 from triggering a full-file scan.
"""

from .model import NoFormatMatch, ParseError

COST_RANK = {"tail": 0, "head": 1, "full-scan": 2}


class FormatSpec:
    __slots__ = ("id", "name", "status", "cost", "extensions", "sniff", "parse", "order")

    def __init__(self, fid, name, status, cost, extensions, sniff, parse, order):
        self.id = fid
        self.name = name
        self.status = status
        self.cost = cost
        self.extensions = extensions
        self.sniff = sniff
        self.parse = parse
        self.order = order


_REGISTRY = []


def register(fid, name, status, cost, extensions, sniff, parse):
    _REGISTRY.append(
        FormatSpec(fid, name, status, cost, extensions, sniff, parse, len(_REGISTRY))
    )


def formats():
    return list(_REGISTRY)


def by_id(fid):
    for f in _REGISTRY:
        if f.id == fid:
            return f
    return None


def sniff_all(reader, opts):
    """Every format's score, in registration order. Used by `dashgps inspect`."""
    out = []
    for spec in _REGISTRY:
        try:
            score = spec.sniff(reader, opts)
        except Exception:
            score = 0.0
        out.append((spec.id, score))
    return out


def _best(cands):
    # Ties break by score, then cheaper IO cost, then registration order. spec 00.
    best = None
    for spec, score in cands:
        key = (-score, COST_RANK[spec.cost], spec.order)
        if best is None or key < best[0]:
            best = (key, spec, score)
    return (best[1], best[2]) if best else (None, 0.0)


def parse_auto(reader, opts, only=None):
    """Detect and parse. Raises NoFormatMatch with every score attached."""
    opts.probe = {}
    if only:
        spec = by_id(only)
        if spec is None:
            raise ParseError("unknown format id: %s" % only)
        return spec.parse(reader, opts)

    scores = []
    for tier, threshold in (("tail", 0.9), ("head", 0.9), ("full-scan", 0.5)):
        if tier == "full-scan" and not opts.deep:
            continue
        cands = []
        confident = None
        for spec in _REGISTRY:
            if spec.cost != tier:
                continue
            try:
                score = spec.sniff(reader, opts)
            except Exception:
                score = 0.0
            scores.append((spec.id, score))
            if score > 0.0:
                cands.append((spec, score))
            # Short-circuit: a confident hit stops the tier, so we never run an expensive
            # scan for a format we are not going to choose. Registration order decides.
            if score >= 0.9:
                confident = spec
                break
        if confident is not None:
            return confident.parse(reader, opts)
        spec, score = _best(cands)
        if spec is not None and score >= threshold:
            return spec.parse(reader, opts)
    raise NoFormatMatch(scores)

File: js/src/registry.js
// Format registry and auto-detection. spec/00-model.md
// Mirrors python/src/dashgps/registry.py

import { NoFormatMatch, ParseError } from "./model.js";

export const COST_RANK = { tail: 0, head: 1, "full-scan": 2 };

const REGISTRY = [];

export function register(id, name, status, cost, extensions, sniff, parse) {
  REGISTRY.push({ id, name, status, cost, extensions, sniff, parse, order: REGISTRY.length });
}

export function formats() { return REGISTRY.slice(); }

export function byId(id) {
  for (const f of REGISTRY) if (f.id === id) return f;
  return null;
}

export async function sniffAll(reader, opts) {
  const out = [];
  for (const spec of REGISTRY) {
    let score = 0;
    try { score = await spec.sniff(reader, opts); } catch (e) { score = 0; }
    out.push([spec.id, score]);
  }
  return out;
}

function best(cands) {
  let win = null;
  for (const [spec, score] of cands) {
    const key = [-score, COST_RANK[spec.cost], spec.order];
    if (win === null || key[0] < win.key[0] ||
        (key[0] === win.key[0] && (key[1] < win.key[1] ||
        (key[1] === win.key[1] && key[2] < win.key[2])))) {
      win = { key, spec, score };
    }
  }
  return win ? [win.spec, win.score] : [null, 0];
}

export async function parseAuto(reader, opts, only) {
  opts.probe = {};
  if (only) {
    const spec = byId(only);
    if (spec === null) throw new ParseError("unknown format id: " + only);
    return spec.parse(reader, opts);
  }
  const scores = [];
  for (const [tier, threshold] of [["tail", 0.9], ["head", 0.9], ["full-scan", 0.5]]) {
    if (tier === "full-scan" && !opts.deep) continue;
    const cands = [];
    let confident = null;
    for (const spec of REGISTRY) {
      if (spec.cost !== tier) continue;
      let score = 0;
      try { score = await spec.sniff(reader, opts); } catch (e) { score = 0; }
      scores.push([spec.id, score]);
      if (score > 0) cands.push([spec, score]);
      // Short-circuit: a confident hit stops the tier, so we never run an expensive scan for a
      // format we are not going to choose. Registration order decides.
      if (score >= 0.9) { confident = spec; break; }
    }
    if (confident !== null) return confident.parse(reader, opts);
    const [spec, score] = best(cands);
    if (spec !== null && score >= threshold) return spec.parse(reader, opts);
  }
  throw new NoFormatMatch(scores);
}
